src/value.mts

```diff
@@ -22,149 +21,268 @@ import { EnvironmentRecord } from './environment.mjs';
-  }
-  switch (typeof value) {
-    case 'undefined':
+export type Value =
+  | UndefinedValue
+  | NullValue
+  | BooleanValue
+  | StringValue
+  | SymbolValue
+  | NumberValue
+  | BigIntValue
+  | PrivateName
+  | ObjectValue;
+
+/** https://tc39.es/ecma262/#sec-ecmascript-language-types */
+export const Value = (() => {
+  // NOTE: Using IIFE so that the class does not conflict with the type of the same name
+  @callable((_target, _thisArg, [value]) => {
+    if (value === null) {
+      return Value.null;
+    } else if (value === undefined) {
       return Value.undefined;
-    case 'string':
-      return new StringValue(value);
-    case 'number':
-      return new NumberValue(value);
-    case 'bigint':
-      return new BigIntValue(value);
-    case 'function':
-      return CreateBuiltinFunction(value, 0, Value(''), []);
-    default:
-      throw new OutOfRange('new Value', value);
-  }
-}) // @ts-expect-error callable class
-abstract class Value {
-  /** @deprecated Use Value() instead of Value() */
-  constructor(value?: never) {
-    if (new.target !== Value) {
-      return this;
+    } else if (value === true) {
+      return Value.true;
+    } else if (value === false) {
+      return Value.false;
     }
-    return Value(value);
-  }
+    switch (typeof value) {
+      case 'string':
+        return createStringValue(value);
+      case 'number':
+        return createNumberValue(value);
+      case 'bigint':
+        return createBigIntValue(value);
+      default:
+        throw new OutOfRange('new Value', value);
+    }
+  })
+  abstract class Value extends BaseValue {
+  }
+  return Value;
+})() as typeof BaseValue & {
+  <T extends null | undefined | boolean | string | number | bigint>(value: T): // eslint-disable-line @engine262/no-use-in-def
+    T extends null ? NullValue :
+    T extends undefined ? UndefinedValue :
+    T extends boolean ? BooleanValue<T> :
+    T extends string ? StringValue :
+    T extends number ? NumberValue :
+    T extends bigint ? BigIntValue :
+    never;
+};
 
-  static declare readonly null: NullValue;
-  static declare readonly undefined: UndefinedValue;
-  static declare readonly true: BooleanValue;
-  static declare readonly false: BooleanValue;
-}
+/** https://tc39.es/ecma262/#sec-ecmascript-language-types */
+export type PropertyKeyValue =
+  | StringValue
+  | SymbolValue;
 
-export class PrimitiveValue extends Value { }
-export type PropertyKeyValue = StringValue | SymbolValue;
+/** https://tc39.es/ecma262/#sec-ecmascript-language-types */
+export type PrimitiveValue =
+  | UndefinedValue
+  | NullValue
+  | BooleanValue
+  | StringValue
+  | SymbolValue
+  | NumberValue
+  | BigIntValue;
+
+/** https://tc39.es/ecma262/#sec-ecmascript-language-types */
+export const PrimitiveValue = (() => {
+  // NOTE: Using IIFE so that the class does not conflict with the type of the same name
+  abstract class PrimitiveValue extends Value {
+  }
+  return PrimitiveValue;
+})();
```

Comment:
what about this?

```suggestion
export const PrimitiveValue = abstract class PrimitiveValue extends Value {}
```

also, does `instanceof PrimitiveValue` work if it does not resolve to a Class entity?

Reply:
> what about this?

TypeScript does not allow `abstract` on class expressions.

> also, does `instanceof PrimitiveValue` work if it does not resolve to a Class entity?

Yes. TypeScript's `instanceof` check uses the construct signature to resolve the instance type.